### PR TITLE
V1

### LINKED ISSUE #30: Finish 1st version
Make 1st version work

### LINKED ISSUE #34: Add logging & dump
We should have an ability to track process of instruction execution stages by logging it & dumping hart's state

### PR DESCRIPTION
**Pull request by Tako-San**

### CODE REVIEW

File: include/elfloader/elfloader.hh [hunk 1]
```diff
@@ -23,6 +23,10 @@ public:
   Addr getEntryPoint() const;
 
   std::span<const Word> getSection(const std::string &name) const;
+  Addr getSectionAddr(const std::string &name) const;
+
+private:
+  ELFIO::section *getSectionPtr(const std::string &name) const;
```
Comment (derzhavin3016) [inline include/elfloader/elfloader.hh]:
```suggestion
  const ELFIO::section *getSectionPtr(const std::string &name) const;
```

File: tools/simulator/main.cc [hunk 2]
```diff
@@ -1,24 +1,33 @@
...
+
 int main(int argc, char **argv) {
   CLI::App app{"Simulator"};
 
   bool isInfo;
   app.add_flag("-i,--info", isInfo, "Turn on info logging level");
 
+  // input
+  fs::path input{};                                    // input
+  app.add_option("input", input, "input")->required(); // input
```
Comment (derzhavin3016) [inline tools/simulator/main.cc]: I suppose you should provide more descriptive comments (or remove them) and more informative help string to option

**Review by derzhavin3016 — CHANGES_REQUESTED** [review]

File: src/decoder/decoder.py [hunk 3]
```diff
@@ -209,12 +209,23 @@ def gen_hh(filename, yaml_dict):
...
 
+    to_write += "namespace sim {\n"
+    to_write += "enum class OpType {\n"
+    to_write += "    UNKNOWN = 0,\n"
     for inst_name in yaml_dict:
         to_write += f"    {inst_name.upper()},\n"
+    to_write += "};\n\n"
 
-    to_write += "};\n"
+    to_write += "inline std::unordered_map<OpType, std::string> opTypeToString {\n"
```
Comment (derzhavin3016) [inline src/decoder/decoder.py]:
```suggestion
    to_write += "inline std::unordered_map<OpType, std::string_view> opTypeToString {\n"
```

File: src/decoder/decoder.py [hunk 4]
```diff
@@ -209,12 +209,23 @@ def gen_hh(filename, yaml_dict):
     """Function to generate c++ header with enum with instructions"""
 
     to_write = COMMENT
-    to_write += "enum class OpType {\n    UNKNOWN = 0,\n"
+    to_write += "#include <string>\n"
```
Comment (derzhavin3016) [inline src/decoder/decoder.py]:
```suggestion
    to_write += "#include <string_view>\n"
```

File: include/common/inst.hh [hunk 5]
```diff
@@ -16,6 +20,24 @@ struct Instruction final {
 
   OpType type{OpType::UNKNOWN};
   RegVal imm{};
+
+  std::string str() const {
+    std::stringstream ss{};
+
+    ss << std::setw(9) << opTypeToString[type] << " {";
```
Comment (derzhavin3016) [inline include/common/inst.hh]: I think you should consider using `at()` here
```suggestion
    ss << std::setw(9) << opTypeToString.at(type) << " {";
```

File: tools/simulator/main.cc [hunk 6]
```diff
@@ -1,24 +1,54 @@
...
+namespace fs = std::filesystem;
+namespace lvl = spdlog::level;
+
+int main(int argc, char **argv) try {
   CLI::App app{"Simulator"};
 
-  bool isInfo;
-  app.add_flag("-i,--info", isInfo, "Turn on info logging level");
+  lvl::level_enum loggingLevel{};
+  std::map<std::string, lvl::level_enum> map{
```
Comment (derzhavin3016) [inline tools/simulator/main.cc:20]:
```suggestion
  std::map<std::string_view, lvl::level_enum> map{
```

**Review by derzhavin3016 — CHANGES_REQUESTED** [review]

File: tools/simulator/main.cc (hunk 6 above)
Comment (Tako-San) [reply tools/simulator/main.cc:20]: Doesn't work

**Review by derzhavin3016 — APPROVED** [review]
LGTMq+

**Pull request merged**